src/ninshiki_cpp/detector/detector.cpp

```diff
@@ -0,0 +1,168 @@
+  std::string config = static_cast<std::string>(getenv("HOME")) + "/yolo_model/config.cfg";
+  std::string weights = static_cast<std::string>(getenv("HOME")) +
+    "/yolo_model/yolo_weights.weights";
+  net = cv::dnn::readNet(weights, config, "");
+
+  this->gpu = gpu;
+  this->myriad = myriad;
+
+  std::ifstream ifs(file_name.c_str());
+  std::string line;
+  while (std::getline(ifs, line)) {
+    classes.push_back(line);
+  }
+
+  // Set computation method (gpu, myriad, or CPU)
+  if (gpu) {
+    net.setPreferableBackend(cv::dnn::DNN_BACKEND_CUDA);
+    net.setPreferableTarget(cv::dnn::DNN_TARGET_CUDA);
+  } else if (myriad) {
+    net.setPreferableBackend(cv::dnn::DNN_BACKEND_OPENCV);
+    net.setPreferableTarget(cv::dnn::DNN_TARGET_MYRIAD);
+  } else {
+    net.setPreferableBackend(cv::dnn::DNN_BACKEND_OPENCV);
+    net.setPreferableTarget(cv::dnn::DNN_TARGET_CPU);
+  }
+}
+
+void Detector::detection(const cv::Mat & image, float conf_threshold, float nms_threshold)
+{
+  std::vector<cv::String> layer_output = net.getUnconnectedOutLayersNames();
+
+  // Create a 4D blob from a frame
+  static cv::Mat blob;
+  cv::Size input_size = cv::Size(416, 416);
+  cv::dnn::blobFromImage(image, blob, 1.0, input_size, cv::Scalar(), true, false, CV_8U);
+
+  net.setInput(blob, "", 0.00392, cv::Scalar(0, 0, 0, 0));
+  std::vector<cv::Mat> outs;
+  net.forward(outs, layer_output);
+
+  // Get width and height from image
+  width = image.cols;
+  height = image.rows;
+
+  static std::vector<int> out_layers = net.getUnconnectedOutLayers();
+  static std::string out_layer_type = net.getLayer(out_layers[0])->type;
+
+  std::vector<int> class_ids;
+  std::vector<float> confidences;
+  std::vector<cv::Rect> boxes;
+
+  if (out_layer_type == "Region") {
+    for (size_t i = 0; i < outs.size(); ++i) {
+      // Network produces output blob with a shape NxC where N is a number of
+      // detected objects and C is a number of classes + 4 where the first 4
+      // numbers are [center_x, center_y, width, height]
+      float * data = reinterpret_cast<float *>(outs[i].data);
+      for (int j = 0; j < outs[i].rows; ++j, data += outs[i].cols) {
+        cv::Mat scores = outs[i].row(j).colRange(5, outs[i].cols);
+        cv::Point classIdPoint;
+        double confidence;
+        cv::minMaxLoc(scores, 0, &confidence, 0, &classIdPoint);
+        if (confidence > conf_threshold) {
+          int centerX = static_cast<int>((data[0] * image.cols));
+          int centerY = static_cast<int>((data[1] * image.rows));
+          int width = static_cast<int>((data[2] * image.cols));
+          int height = static_cast<int>((data[3] * image.rows));
+          int left = centerX - width / 2;
+          int top = centerY - height / 2;
+
+          class_ids.push_back(classIdPoint.x);
+          confidences.push_back(static_cast<float>(confidence));
+          boxes.push_back(cv::Rect(left, top, width, height));
```

Comment:
instead of using `static_cast` twice (float->int->float) for the boxes, use cv::Rect2f instead for saving these boxes

Reply:
I don't think we can do that because the first parameter `cv::dnn::NMSBoxes` that we need is either `std::vector<Rect>` or `std::vector<Rect2d>`. Therefore, if we use `cv::Rect2f` for boxes we need to static_cast again later on.